hw/net/rocker/rocker_p4_rmt.c

```diff
@@ -0,0 +1,216 @@
+        return;
+    }
+    if (w == NULL) {
+        DPRINTF("P4-RMT world not found\n");
+        return;
+    }
+    // check for sup port
+    if (p4_rmt->is_cpu_port(eg_port1)) {
+        DPRINTF("rocker_p4_rmt_tx to CPU from ingress port %d, len %d.... \n", 
+                    ig_port1, len);
+        rx_produce(w, ig_port1, &iov, iovcnt);
+        return;
+    }
+    // call qemu_send to send out on switch port
+    rocker_port_eg(rocker, eg_port1, &iov, iovcnt);
+    return;
+}
+
+static int rocker_p4_rmt_table_cmd (struct p4_rmt_world *p4_rmt, 
+                            struct desc_info *info,
+                            char *buf, unsigned short cmd,
+                            RockerTlv **tlvs)
+{
+    unsigned int    table_id;
+    unsigned char   *entry;
+    if (tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ID] == NULL ||
+        tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ENTRY] == NULL) {
+        DPRINTF("Missing TLVs\n");
+        return 0;
+    }
+        
+    table_id = rocker_tlv_get_u32(tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ID]);
+    entry = rocker_tlv_data(tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ENTRY]);
+
+    DPRINTF("p4_rmt_table_cmd for table %d\n", table_id);
+
+    if (!p4_rmt->is_table_valid(table_id)) {
+        DPRINTF("p4_rmt_table_cmd for invalide table %d\n", table_id);
+        return 0;
+    }
+
+    switch (cmd) {
+        case ROCKER_TLV_CMD_TYPE_P4_RMT_TABLE_ENTRY_ADD:
+            // check if fn is provided
+            if (p4_rmt->table_ops[table_id].add) {
+                p4_rmt->table_ops[table_id].add(entry);
+            }
+            break;
+        case ROCKER_TLV_CMD_TYPE_P4_RMT_TABLE_ENTRY_MOD:
+        case ROCKER_TLV_CMD_TYPE_P4_RMT_TABLE_ENTRY_DEL:
+            break;
+        case ROCKER_TLV_CMD_TYPE_P4_RMT_TABLE_DEFAULT_ACTION:
+        {
+            unsigned int action_id = rocker_tlv_get_u32(
+                            tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ENTRY]);
+            // XXX - add support for action data for default action
+            unsigned char data[4] = {0,0,0,0}; // HACK
+            if (p4_rmt->table_ops[table_id].default_action) {
+                p4_rmt->table_ops[table_id].default_action((int)action_id, 
+                                                            data);
+            }
+            break;
+        }
+    }
+    return 0;
+}
+
+ssize_t rocker_p4_rmt_ig (World *world, unsigned int pport,
+                            const struct iovec *iov, int iovcnt)
+{
+    int pkt_len = iov_size(iov, iovcnt);
+    unsigned int port = pport;  // model uses pports (1-N)
+    struct p4_rmt_world *p4_rmt = world_private(world);
+
+    // ingress pipeline
+    DPRINTF("p4_rmt_ingress port %d, len %d .... \n", port, pkt_len);
+
+    // convert iov to a pkt buffer
+    unsigned char *buf = rocker_p4_iov_to_buffer(iov, iovcnt);
+    if (buf == NULL) {
+        return 0;
+    }
+    p4_rmt->process_pkt(port, buf, pkt_len);
+    return iov_size(iov, iovcnt);
```

Comment:
By the model code. (library)

On Sun, Apr 19, 2015 at 10:33 PM, Scott Feldman <notifications@github.com>
wrote:

> In hw/net/rocker/rocker_p4_rmt.c
> <https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28664806>:
>
> > +                            const struct iovec *iov, int iovcnt)
> > +{
> > +    int pkt_len = iov_size(iov, iovcnt);
> > +    unsigned int port = pport;  // model uses pports (1-N)
> > +    struct p4_rmt_world *p4_rmt = world_private(world);
> > +
> > +    // ingress pipeline
> > +    DPRINTF("p4_rmt_ingress port %d, len %d .... \n", port, pkt_len);
> > +
> > +    // convert iov to a pkt buffer
> > +    unsigned char *buf = rocker_p4_iov_to_buffer(iov, iovcnt);
> > +    if (buf == NULL) {
> > +        return 0;
> > +    }
> > +    p4_rmt->process_pkt(port, buf, pkt_len);
> > +    return iov_size(iov, iovcnt);
>
> Where does buf get freed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pbhide/qemu-rocker/pull/1/files#r28664806>.
>